### ISSUE TITLE
Support to Preact X

### ISSUE THREAD

**Issue by fjorgemota** [prateekbh/preact-material-components#1301]
Hey.

I observed that there's no support to Preact X yet.

Specifically, the typescript compiler emits an error about code using the `attributes` attribute in the VNode structure, like [this example](https://github.com/prateekbh/preact-material-components/blob/master/ts/Base/MaterialComponent.ts#L170-L178).

I imagine this is related to the fact that in Preact X the VNode shape has changed, and now `attributes` is just `props` in the VNode structure. According to Typescript compiler, there's a few other cases where the same error occurs.

So here's the question: What's the plan to support Preact X? Would this package have a version compatible with both versions (assuming this is the unique problem, of course), or we would release a new version supporting only Preact X?

PS: I know that Preact X is still in alpha but I think it's better to prepare for it than await to a stable release...What you think?

**Comment by cromefire** [comment]
@prateekbh it seems not that big of a change (Read through the breaking changes section), should we put it in `2.0`?

**Comment by cromefire** [comment]
@fjorgemota could you try this against 2.0 (it's in the `2.0` branch of the repo)?

**Comment by cromefire** [comment]
> Would this package have a version compatible with both versions

Yes

**Comment by fjorgemota** [comment]
Hey @cromefire, I didn't tested the version 2.0 here because it seems to not be a unique package anymore and I honestly don't know how to test it here on my project.

But..[this](https://github.com/prateekbh/preact-material-components/blob/0e2e79766039e4f68cf723cef053316cf4284b3b/packages/base/src/MaterialComponent.ts#L189-L198) shows that the 2.0 branch has the same problem as master problem, and would not be compatible with Preact X in the actual state.

> > Would this package have a version compatible with both versions
>
> Yes

Well, I think it would be good if we have some helpers (probably in the base package) to manage the differences in the VNode structure, then...specifically the `attributes` (now `props`) attribute. Depending on the number of changes involved, of course, as the overhead over users using Preact X directly may grow overtime (with more breaking changes here and there over the Preact 8 version), I think. Just my two cents. :)

**Comment by johnhaitas** [comment]
@fjorgemota I have an experimental branch where i made the basic vnode shape changes for Preact X ... It seems that preact 10.0.0-alpha.2 has a bug that has already been fixed on the master branch so i had to link the preact package locally. Preact 10.0.0-alpha.3 should fix that though. There are still some problems I noticed. For example, the `IconButton` class has a `on` prop that throws an error with devtools because `on` is expected to be a function rather than a boolean. Here is my diff with `preact-material-components` `master` branch

https://github.com/prateekbh/preact-material-components/compare/master...johnhaitas:experimental/preact-10

**Comment by cromefire** [comment]
@johnhaitas 1.X is in maintenance mode it will only get bug fixes, for something new please develop against `2.0`

**Comment by fjorgemota** [comment]
Hey @johnhaitas! Nice that you started working on the changes related to add support to Preact X to preact-material-components. :D 

As @cromefire noted, these changes should be applied against 2.0 to be applied to the next version...And also the support to Preact 8 shouldn't be removed as commented by @cromefire [here](https://github.com/prateekbh/preact-material-components/issues/1301#issuecomment-476402979).

By the way, great that you're using `toChildArray` from Preact X to get the same array as Preact 8. It's another thing that'll be needed to support Preact X, together with VNode shape changes.

**Comment by Drapegnik** [comment]
Hi! If there are any updates with this?

**Comment by prateekbh** [comment]
Hi, we are pretty close to our 2.0 alpha with just tabs remaining
we'll fix this post alpha and also send a fix to 1.x next week

**Comment by Drapegnik** [comment]
## with ts:
![Screen Shot 2019-09-03 at 17 34 01](https://user-images.githubusercontent.com/11758660/64182809-78832e00-ce71-11e9-9966-83fb7d72aa80.png)

```bash
node_modules/preact-material-components/Base/MaterialComponent.d.ts:10:105 - error TS2503: Cannot find namespace 'JSX'.

10 export declare type MaterialComponentProps<PropType> = SoftMerge<PropType & IMaterialComponentOwnProps, JSX.HTMLAttributes>;
                                                                                                           ~~~

node_modules/preact-material-components/Base/MaterialComponent.d.ts:25:19 - error TS2314: Generic type 'MDCComponent<FoundationType>' requires 1 type argument(s).

25     MDComponent?: MDCComponent<any, any>;
                     ~~~~~~~~~~~~~~~~~~~~~~

node_modules/preact-material-components/Button/index.d.ts:18:41 - error TS2503: Cannot find namespace 'JSX'.

18 export declare class Button<PropsType = JSX.HTMLAttributes, StateType = {}> extends MaterialComponent<PropsType & IButtonProps, StateType & IButtonState> {

                                           ~~~

node_modules/preact-material-components/Button/index.d.ts:22:40 - error TS2503: Cannot find namespace 'JSX'.

22     protected materialDom(props: any): JSX.Element;
                                          ~~~

node_modules/preact-material-components/Icon/index.d.ts:9:40 - error TS2503: Cannot find namespace 'JSX'.

9     protected materialDom(props: any): JSX.Element;
                                         ~~~

[5:31:38 PM] Found 5 errors. Watching for file changes.
```

**Comment by cromefire** [comment]
> error TS2503: Cannot find namespace 'JSX'

For the `JSX` namespace you must have misconfigured something with the preact types. I think you have to configure the _react namespace_ to `preact` in ts.

**Comment by Drapegnik** [comment]
> 1.5.8 • Public • Published 3 months ago

hmm, does `1.5.8` inludes `Preact X` fix?

@prateekbh, how could I test it?

**Comment by cromefire** [comment]
> > 1.5.8 • Public • Published 3 months ago
> 
> hmm, does `1.5.8` inludes `Preact X` fix?

The Preact X fix is less than 3 Month old

**Comment by Drapegnik** [comment]
@cromefire, https://github.com/preactjs/preact/issues/1036#issuecomment-445890450
I already have
```json
    "jsx": "react",
    "jsxFactory": "h",
```
in my config and it works.

but falls with `preact-material-components`

**Comment by Drapegnik** [comment]
> The Preact X fix is less than 3 Month old

so, is it published? can it be used?

**Comment by cromefire** [comment]
Okay it seems like one would have to alias `preact.JSX` as `JSX` which we can't do because it would result in incompatibilities with `8.X`

Something like `globals.d.ts`:
```typescript
import export JSX = preact.JSX;
```
could do it.

But I did not test it.

**Comment by prateekbh** [comment]
i'll release it today

**Comment by prateekbh** [comment]
sorry moving this release to end of week, just need to be doubly sure before release

**Comment by prateekbh** [comment]
now finding how to change channel from @next to stable

**Comment by prateekbh** [comment]
released 1.6.0

**Comment by Drapegnik** [comment]
I confirm that works with:

```tsx
declare global {
  // FIXME
  // https://github.com/preactjs/preact/issues/1036#issuecomment-491409239
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    type Element = JSXInternal.Element;
    type HTMLAttributes = JSXInternal.HTMLAttributes;
  }
}
```

- `preact-material-components@1.6.0`
- `preact@10.0.0-rc.2`
